### DISCUSSION TITLE
Handle the Enter key?

### DISCUSSION THREAD

**Discussion by kwackerz** [tui-cs/Terminal.Gui#2692] [General]
I am struggling to handle the enter key in my app.
I have a barcode scanner that emulates a keyboard
the window is handling the keypresses and has no text fields that can 'grab' the input
I also have 'back' button

when the window receives the Enter from the scanner ( at the end of the barcode) I can se it in window key events
but after I handle it the btnBack.Clicked event fires. even if I set btnBack.Canfocus = false;

I know I can change the scanner hardware to use a different suffix but I don't want to have to do this.
Any idea how I can handle this situation?

**Comment by BDisp** [comment]
When you handle the `Key.Enter` in the `Window`, I think you are using `win.KeyPress` event, you have to set the `Handled` property to true to avoid the `btnBack` processing the key.

**Reply by kwackerz** [reply]
the Enter key does not seem to fire the keypress event. I can get it via KeyDown or KeyUp but args.Handled = true doesn't stop it from triggering the button.clicked.

**Reply by kwackerz** [reply]
ok, so.... I think what is happening is the enter key is being processed by a tableview on the previous window. something else I don't quite understand... YET :-)

**Reply by BDisp** [reply]
If you have subscribed the `KeyPress` into another views then the first one will process it on the same `Application.Current`. You must provide some code that can reproduce the same behavior as you are facing.

**Reply by BDisp** [reply]
I already understand ed what it's happening. Since the back button has the focus when you press the Enter key the button will be processed. The trick is the Window subscribe the `KeyDown` and the Button subscribe the `KeyPress`. Here is a sample that works:

```cs
			Application.Init ();
			var win = new Window ();
			var btnBarcode = new Button ("Barcode >") { X = Pos.AnchorEnd (13) };
			btnBarcode.Clicked += (s, e) => {
				var win2 = new Window ();
				win2.KeyDown += (s, e) => {
					if (e.KeyEvent.Key == Key.Enter) {
						MessageBox.Query ("Barcode", "Register processed!", "Ok");
						e.Handled = true;
					}
				};
				var btnBack = new Button ("< Back");
				btnBack.KeyPress += (s, e) => {
					if (e.KeyEvent.Key == Key.Enter) {
						e.Handled = true;
					}
				};
				btnBack.Clicked += (s, e) => Application.RequestStop ();
				win2.Add (btnBack);
				Application.Run (win2);
			};
			win.Add (btnBarcode);
			Application.Top.Add (win);

			Application.Run ();
			Application.Shutdown ();

```

**Reply by kwackerz** [reply]
Perfect BDisp. Thank you so much!
